tools/generate_type_hints.py

```diff
@@ -0,0 +1,221 @@
+    _UInt32CodesBase,
+    _UInt64CodesBase,
+    _Float16CodesBase,
+    _Float32CodesBase,
+    _Float64CodesBase,
+    _Complex64CodesBase,
+    _Complex128CodesBase,
+    _ByteCodes,
+    _ShortCodes,
+    _IntCCodes,
+    _IntPCodes,
+    _IntCodes,
+    _LongLongCodes,
+    _UByteCodes,
+    _UShortCodes,
+    _UIntCCodes,
+    _UIntPCodes,
+    _UIntCodes,
+    _ULongLongCodes,
+    _HalfCodes,
+    _SingleCodes,
+    _DoubleCodes,
+    _LongFloatCodes,
+    _CSingleCodes,
+    _CDoubleCodes,
+    _CLongFloatCodes,
+)
+
+_Int8Codes = Union[_Int8CodesBase{int8}]
+_Int16Codes = Union[_Int16CodesBase{int16}]
+_Int32Codes = Union[_Int32CodesBase{int32}]
+_Int64Codes = Union[_Int64CodesBase{int64}]
+
+_UInt8Codes = Union[_UInt8CodesBase{uint8}]
+_UInt16Codes = Union[_UInt16CodesBase{uint16}]
+_UInt32Codes = Union[_UInt32CodesBase{uint32}]
+_UInt64Codes = Union[_UInt64CodesBase{uint64}]
+
+_Float16Codes = Union[_Float16CodesBase{float16}]
+_Float32Codes = Union[_Float32CodesBase{float32}]
+_Float64Codes = Union[_Float64CodesBase{float64}]
+
+_Complex64Codes = Union[_Complex64CodesBase{complex64}]
+_Complex128Codes = Union[_Complex128CodesBase{complex128}]
+
+# Note that these variables are private despite the lack of underscore;
+# don't expose them to the `numpy.typing` __init__ file.
+byte = {byte}
+short = {short}
+intc = {intc}
+intp = int0 = {intp}
+int_ = {int_}
+longlong = {longlong}
+
+ubyte = {ubyte}
+ushort = {ushort}
+uintc = {uintc}
+uintp = uint0 = {uintp}
+uint = {uint}
+ulonglong = {ulonglong}
+
+half = {half}
+single = {single}
+float_ = double = {float_}
+longfloat = longdouble = {longfloat}
+
+csingle = singlecomplex = {csingle}
+complex_ = cdouble = cfloat = {complex_}
+clongfloat = longcomplex = clongdouble = {clongfloat}
+"""
+
+
+def _file_to_context(
+    file: Union[IO[str], _AnyPath], *args: Any, **kwargs: Any
+) -> ContextManager[IO[str]]:
+    """Create a context manager from a path- or file-like object."""
+    try:
+        return open(file, *args, **kwargs)
+    except TypeError:
+        return nullcontext(file)
+
+
+def generate_alias(file: Union[IO[str], _AnyPath] = sys.stdout) -> None:
+    """Generate a stub file with platform-specific types.
+
+    Parameters
+    ----------
+    file : file- or path-like
+        A file- or path-like object used for writing the results.
+
+    """
+    # Generate type aliases and update the character codes for all known types
+    type_alias = {}
+    char_codes: Dict[str, str] = defaultdict(str)
+
+    for name, code_names in TO_BE_ANNOTATED.items():
+        typ = getattr(nt, name)
+        if typ in ANNOTATED:
+            type_alias[name] = typ.__name__
+            char_codes[typ.__name__] += f", {code_names}"
+        else:
+            # Plan B in case an untyped class is encountered
+            type_alias[name] = "Any"
+
+    # Generate the docstring
+    now = dt.datetime.now()
+    docstring = f'"""THIS FILE WAS AUTOMATICALLY GENERATED ON {now}."""'
```

Comment:
I think we've had complaints in the past about unreproducible builds caused by putting timestamps in things, so I'd lean towards not doing so here.

Reply:
Done as of https://github.com/numpy/numpy/pull/17514/commits/5887e625eac696d03299bcc926cd1b5c3d7d5a3a.